### PR TITLE
Ensure locks are released in error scenario

### PR DESCRIPTION
**Pull request by quantumagi**
https://stratisplatformuk.visualstudio.com/StratisBitcoinFullNode/_workitems/edit/4589

### CODE REVIEW

File: src/Stratis.Features.SQLiteWalletRepository/Tables/HDWallet.cs [hunk 1]
```diff
@@ -111,7 +111,7 @@ internal class HeightHashPair
             internal string Hash { get; set; }
         }
 
-        internal static void AdvanceTip(SQLiteConnection conn, HDWallet wallet, ChainedHeader newTip, uint256 prevTipHash)
+        internal static void AdvanceTip(DBConnection conn, HDWallet wallet, ChainedHeader newTip, uint256 prevTipHash)
```
Comment (fassadlr) [inline src/Stratis.Features.SQLiteWalletRepository/Tables/HDWallet.cs]: How come this change?

File: src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs [hunk 2]
```diff
@@ -882,17 +882,21 @@ private bool ParallelProcessBlock(ProcessBlocksInfo round, Block block, ChainedH
 
                                 conn.Rollback();
 
+                                throw;
```
Comment (fassadlr) [inline src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs]: These changes appears to be unnecessary. Previously, we would end batch in the catch (which catches all exceptions) so moving it to the `finally` doesn't really add any value? Unless it is moved there to ensure that it always calls end batch in non-error scenarios? Or is that what the extra `else` is for further down?

Unless I'm missing something here 😎

File: src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs [hunk 3]
```diff
@@ -882,17 +882,21 @@ private bool ParallelProcessBlock(ProcessBlocksInfo round, Block block, ChainedH
...
+                                this.Metrics.ProcessTime += (DateTime.Now.Ticks - flagFall);
+                                this.Metrics.LogMetrics(this, conn, chainedHeader, wallet);
                             }
-
-                            this.Metrics.ProcessTime += (DateTime.Now.Ticks - flagFall);
-                            this.Metrics.LogMetrics(this, conn, chainedHeader, wallet);
                         }
-
-                        this.EndBatch(round);
+                        else
```
Comment (fassadlr) [inline src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs]: What was the purpose of adding an `else` here?

File: src/Stratis.Features.SQLiteWalletRepository/Tables/HDWallet.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Features.SQLiteWalletRepository/Tables/HDWallet.cs]: After the change the `DBConnection.Execute` method, which includes logging, will be used.

File: src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs]: The idea is to always call `EndBatch` where errors in `Rollback` could previously prevent that.

File: src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs]: To avoid `EndBatch` being called twice.

**Comment by quantumagi** [comment]
I am not sure what the actual problem is but these changes will achieve 2 things:
1) Prevent the wallet sync from stopping or failing due to inability to acquire non-released locks.
2) Add more logging to help determine any remaining issue.

File: src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepository.cs]: Ah ok yes, cool :)

**Pull request merged**